### PR TITLE
[longhaul] Fixed memory leaks and thread locks

### PR DESCRIPTION
**Pull request by ericwol-msft**
# Checklist
- [x] I have read the [contribution guidelines] (https://github.com/Azure/azure-iot-sdk-c/blob/master/.github/CONTRIBUTING.md).
- [x] I added or modified the existing tests to cover the change (we do not allow our test coverage to go down).
- If this is a modification that impacts the behavior of a public API
  - [ ] I edited the corresponding document in the `devdoc` folder and added or modified requirements.
- I submitted this PR against the correct branch: 
  - [x] This pull-request is submitted against the `master` branch. 
  - [x] I have merged the latest `master` branch prior to submission and re-merged as needed after I took any feedback.
  - [x] I have squashed my changes into one with a clear description of the change.

# Reference/Link to the issue solved with this PR (if any)

# Description of the problem

# Description of the solution

### CODE REVIEW

File: testtools/iothub_test/src/iothub_account.c [hunk 1]
```diff
@@ -749,6 +757,7 @@ IOTHUB_ACCOUNT_INFO_HANDLE IoTHubAccount_Init_With_Config(IOTHUB_ACCOUNT_CONFIG*
                     if (retrieveConnStringInfo(iothub_account_info) != 0)
                     {
                         LogError("retrieveConnStringInfo failed.\r\n");
+                        free(iothub_account_info->sasDevices);
```
Comment (ewertons) [inline testtools/iothub_test/src/iothub_account.c:760]: Aren't x509Certificate and x509PrivateKey also supposed to be freed in all the error blocks below?

File: testtools/iothub_test/src/iothubtest.c [hunk 2]
```diff
@@ -1158,8 +1195,13 @@ IOTHUB_TEST_CLIENT_RESULT IoTHubTest_ListenForEventAsync(IOTHUB_TEST_HANDLE devh
...
             }
 
+            if (devhubValInfo->lock != NULL)
+            {
+                Lock_Deinit(devhubValInfo->lock);
+            }
+
             devhubValInfo->onMessageReceivedCallback = NULL;
-            devhubValInfo->onMessageReceivedContext = NULL;
+            devhubValInfo->onMessageReceivedContext = NULL;
```
Comment (ewertons) [inline testtools/iothub_test/src/iothubtest.c:1204]:
```suggestion
            devhubValInfo->onMessageReceivedContext = NULL;
```

File: testtools/iothub_test/src/iothubtest.c [hunk 3]
```diff
@@ -1170,6 +1212,11 @@ IOTHUB_TEST_CLIENT_RESULT IoTHubTest_ListenForEventAsync(IOTHUB_TEST_HANDLE devh
                 LogError("Already listening for messages");
                 result = IOTHUB_TEST_CLIENT_ERROR;
             }
+            else if ((devhubValInfo->lock = Lock_Init()) == NULL)
```
Comment (ewertons) [inline testtools/iothub_test/src/iothubtest.c:1215]: Where is this lock being destroyed?

**Review by ewertons — APPROVED** [review]

File: testtools/iothub_test/src/iothubtest.c (hunk 3 above)
Comment (ericwol-msft) [reply testtools/iothub_test/src/iothubtest.c:1215]: line 1200

**Pull request merged**